### ISSUE TITLE
Board view logic improvement

### ISSUE THREAD

**Issue by CloneWith** [CloneWith/osu#10]
- [x] Rewrite the beatmap listing logic (manually specified)
- [x] Built-in board editor (considering)

---

Various problems have found in OFFC Test Match:

- [x] When A trap is placed, `InstructionDisplay` won't display the correct trap information
  - *Found in the latest build in `derived/tboard`, probably caused by unexpected/unpredicted box status*

- [x] Two messages maybe displayed at the same time (when EX will be triggered at the same time)
  - Can be recreated using following steps:
    1. Set `大陆漂移` trap
    2. Trigger this trap using pick
    3. Use `Perform Swap`
  - ~即将漂移EX模式~. See this issue #18

**Comment by NaughtyChas** [comment]
> * [ ]  Rewrite the beatmap list logic (manually specified)

Beatmap listing can be achieved by edit beatmaps in correct order inside `Rounds Editor`:

|Rounds Editor|Board|
|:-:|:-:|
|![image](https://github.com/user-attachments/assets/7089b59a-97f7-437a-9225-9079581a646c)|![image](https://github.com/user-attachments/assets/00efdb2f-1b25-4b1d-97db-ea6f7860315c)|

Listing from top to bottom in `Rounds Editor` will result in a expected order from top to bottom, from left to right.

This issue is worth to be re-thinked.

**Comment by CloneWith** [comment]
Not until now have I reliazed this trick!
So I think we temporarily don't need a Board-specific editor?

**Comment by NaughtyChas** [comment]
> Not until now have I reliazed this trick! So I think we temporarily don't need a Board-specific editor?

Depends on display logic of EX Screen. The board should in 4x4 size, adding more than 16 beatmaps will cause board display issue, as shown in the pic below:

![image](https://github.com/user-attachments/assets/b2233d40-1eb1-4e68-b2de-2a94cfcf81d8)

In this case I have added 16 + 4 beatmaps in current round, for 16 maps in board & 4 EX maps. The board is now in a 5x5 display which is not we are expecting.

---

Here are some potential solution towards this issue:

1. Change board's beatmap display logic to make it **not render beatmaps labelled as EX**.
2. Add a EX editor.

It is obvious that the second solution should be easier to implement. Problem is it will cause some degree of redundant if we add a dedicated tab on the left. A suggested approach is to **add four input boxes on the right of the EX screen**, each of them accept beatmap IDs, for EX 1-4 respectively.

**Comment by CloneWith** [comment]
I think we've somehow implemented solution 1 in the EX Stage view, where we only add `mod=EX` beatmaps to the list, refer to [this commit](https://github.com/CloneWith/osu/commit/59776105e19026e78b27ab0ef3497f9e52e300b1).

Noticeably a disadvantage of this implementation is *portability*, which means we can't use other mod names other than `EX` to make a map visible in EX view. Chances are that I'll add this as a JSON argument and enable it to be configured in the main settings screen.

**Comment by NaughtyChas** [comment]
> I think we've somehow implemented solution 1 in the EX Stage view, where we only add `mod=EX` beatmaps to the list, refer to [this commit](https://github.com/CloneWith/osu/commit/59776105e19026e78b27ab0ef3497f9e52e300b1).
> 
> Noticeably a disadvantage of this implementation is _portability_, which means we can't use other mod names other than `EX` to make a map visible in EX view. Chances are that I'll add this as a JSON argument and enable it to be configured in the main settings screen.

Solution 1 is indeed already implemented in the code, and it works well in the perspective of EX view. However adding 4 extra EX maps in Round Editor, with `Mods` defined as `EX` with blank `Mod Index`, these maps somehow will display in the board, which is not expected.

|Round Editor||
|:-:|:-:|
|![image](https://github.com/user-attachments/assets/003a6ef1-d7a6-4f63-82a1-c020e7951cb9)|![image](https://github.com/user-attachments/assets/ec189744-5191-46ea-9d57-fd04dd262661)|

The first solution can be marked as complete only when these EX Maps are not being rendered on the board view.

**Comment by CloneWith** [comment]
> The first solution can be marked as complete only when these EX Maps are not being rendered on the board view.

Fixed in the latest commit.

| Map Settings | Board View |
| :-: | :-: |
| ![Settings](https://github.com/user-attachments/assets/81a487cf-e002-4642-9556-ea1a4823f663) | ![Board](https://github.com/user-attachments/assets/64f112b7-a513-4781-95f6-0a1acdbde53f) |

**Comment by NaughtyChas** [comment]
> > The first solution can be marked as complete only when these EX Maps are not being rendered on the board view.
> 
> Fixed in the latest commit.
> 
> Map Settings	Board View
> ![Settings](https://github.com/user-attachments/assets/81a487cf-e002-4642-9556-ea1a4823f663)	![Board](https://github.com/user-attachments/assets/64f112b7-a513-4781-95f6-0a1acdbde53f)

Reviewed, cheers for your fantastic work! This issue can be drown to a close, since we have addressed how we're going to fill the board with corresponding beatmaps.

**Comment by NaughtyChas** [comment]
commit [bc3507f](https://github.com/CloneWith/osu/commit/bc3507fcd8fd2a3d0ce17bd8e67e95a59ddb557a) added board arrangement edit based on Board row/column. only input boxes are being created so far:

![image](https://github.com/user-attachments/assets/fd0e3b32-7f96-4ae9-bd9d-7f635df5cc4d)

---

> * [x]  Rewrite the beatmap list logic (manually specified)

Re-opening first checkbox, as no specific methods is being declared sofar to put correspond beatmap into boxes. Re-check it if we complete this.

---

As we have to give every beatmap a position, automatic EX situation judgement should be easier to get implemented. Opening new issue at

**Comment by CloneWith** [comment]
> Re-opening this issue and the first task, as no specific methods is being declared so far to put correspond beatmap into boxes.

This is implemented in our recent commits.

![The switch](https://github.com/user-attachments/assets/4256289c-b4ae-4595-a37e-501c85801d63)

To prevent boards being broken by maps with undefined board coodinates, I've added a switch in the header of each round. To use the manually set position, we just need to turn the switch on and then it will work.

**Comment by NaughtyChas** [comment]
@CloneWith are there any thing left we haven't finish in this issue? Closing this issue if ready.

**Comment by CloneWith** [comment]
Seemingly most stuff about this has been down, and we can safely close this now.
